Enrich unit tests for nested Array data type

B-Step62 opened the pull request:
### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

`Array` data type has been added to support LLM input (#9893). This PR add test cases for nested Array.

A few changes are made for a few edge cases found;
1. Add recursive call in `_enforce_array` to support nested array.
2. Remove conversion from single scalar to list when specified data type is `Array`, at [models/utils.py#L910-L911](https://github.com/mlflow/mlflow/blob/llm_signature/mlflow/models/utils.py#L910-L911). IMO we should throw exception when data type is Array but value is not.

**Callout**
I've also found an interesting case with an empty array. In current implementation of [_infer_colspec_type](https://github.com/mlflow/mlflow/blob/llm_signature/mlflow/types/utils.py#L106), we don't allow empty list, which is reasonable as we can't infer schema from it. However, this also throws exception when nested array has empty sub-array, like `[["a", "b"], []]`, as it simply call `_infer_colspec_type` to each element recursively.

I'm wondering if we want to allow this case, because we can infer column type using the other element in this case. Probably need concrete example usage to conclude... idk what kind of scenario model input has nested array, and whether or not it's likely to have empty sub-array. Interested to hear opinions from other ppl.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

B-Step62 committed:
Enhance utility test for nested Array

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
refactor

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Add nested array test case to pyfunc_schema_enforcement

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

serena-ruan commented:
> However, this also throws exception when nested array has empty sub-array, like [["a", "b"], []], as it simply call _infer_colspec_type to each element recursively.

I think in this case the second element is invalid. If the column is optional, you should provide none/nan instead of an empty list.

serena-ruan commented on tests/models/test_utils.py:
```diff
@@ -321,36 +321,80 @@ def test_enforce_property_with_errors():
+            ),
+            Array(Array(DataType.string)),
+        ),
+        # 6. Empty array
+        (
+            [],
+            Array(DataType.string),
+        ),
```
I think this should fail, let's add a check in _enforce_array function.

serena-ruan commented on tests/models/test_utils.py:
```diff
@@ -321,36 +321,80 @@ def test_enforce_property_with_errors():
+        ),
+    ],
+)
+def test_enforce_array(data, schema):
```
Can we add some tests where dtype is not DataType? I'm wondering if nested numpy arrays have any problems.

B-Step62 replied:
Sure, I guess it works fine but I'll make it sure.

B-Step62 replied:
Note from offline: With current behavior,`np.array` passed is returned as python list. This is not a bug but ideally we don't make any side effect to the given data. Will work on this in separate follow-up PR.

serena-ruan commented:
@yukimori This looks good! Besides array enforcement, could we add tests for schema inference as well? Specifically`_infer_schema` or `_infer_signature` and test them in a PyFuncModel to make sure e2e works.

B-Step62 commented:
> This looks good! Besides array enforcement, could we add tests for schema inference as well? Specifically`_infer_schema` or `_infer_signature` and test them in a PyFuncModel to make sure e2e works.

@serena-ruan Gotcha. Added a case to pyfunc (`test_pyfunc_model_schema_enforcement_complex`) for e2e check, but will add a few more for those inference methods.

(btw you tagged wrong Yuki haha:p)

B-Step62 committed:
Add check for empty list

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Add more tests on infer_schema / infer_signature

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 commented on mlflow/models/utils.py:
```diff
@@ -906,15 +906,21 @@ def _enforce_datatype(data: Any, dtype: DataType):
     if not required and data is None:
         return None
+
+    if len(data) == 0:
+        raise MlflowException(
+            "Expected a non-empty list/array for Array column. If you want an empty value, "
+            "mark the column `required=False` and put `None` instead."
+        )
```
Here added check for empty list/array. `len()` doesn't catch nested list like `[[], []]` but it will be caught in recursive call anyway.

harupy replied:
What would happen if we accept an empty list/array?

B-Step62 replied:
Not an issue of enforcing given schema on data indeed, while it causes inconsistency with `_infer_colspec_type`, which doesn't allow empty array (for inner-element of nested array as well).

However, on second thought, we should probably be able to allow empty array here. As far as I see [_infer_colspec_type code](https://github.com/mlflow/mlflow/blob/llm_signature/mlflow/types/utils.py#L121-L124), it should only block empty array at top level (sample level), rather than element in each sub-list, e.g. `{"data": []}` is not allowed but `{"data: [["a", "b"],  ["c", "d"], []]}` can be allowed, because we can infer the column type from the other elements. If we make this change, removing chech in `_enforce_array` won't cause inconsistency.

WDYT @harupy @serena-ruan ?

harupy replied:
Thanks for the explanation! 

1. I agree. They don't have to be the same. `_enforce_array` should accept an empty array. The model should decide what to do with it.
2. I agree. `_infer_colspec_type code` should work for `{"data": [["a", "b], []}` (unless it's super complex to implement it).

B-Step62 replied:
Awesome! Will include fix for (1) in this PR and cut a separate one for (2)

serena-ruan replied:
I don't think `_infer_colspec_type` should work for `{"data": [["a", "b], []]}`. It should be inferred as `Array(Array(string))`, and we have the assumption that elements in an array should be of same type. I think it could be None like `{"data": [["a", "b"], None, np.nan]}` for `ColSpec(Array(Array(string)), required=False)`, but empty list is not omitting the value.

serena-ruan replied:
For 1: from my opinion I think it shouldn't pass for `_enforce_array` on an empty array, but let's check what's the behavior for 2.7.1 and keep it.

B-Step62 replied:
IMO empty list is different from None, the former represents no element while the latter is null.

For example, let's say there is a column listing name of students in a club
```
{"list_of_students": []}
```
says the club doesn't have any member. However,
```
{"list_of_students": None}
```
says no info on the column -> optional.

B-Step62 replied:
@harupy @serena-ruan 
Thank you for providing your thoughts! It seems both perspective have some legit points, but I have to get a consensus to close this PR;

For 1, apparently [_enforce_mlflow_datatype](https://github.com/mlflow/mlflow/blame/v2.7.1/mlflow/models/utils.py#L382-L388) in 2.7.1 does nothing for an empty list or a nested list includes an empty list, as it basically avoid touching object type much. I'm not 100% sure and please correct me if I'm wrong, but if this is the case, can we align on allowing empty list here?

For 2, I think the key point is whether we are ok with this assumption

> we have the assumption that elements in an array should be of same type.

I think assuming type based on non-empty values is fine, and we can still throw exception when the inputs are all empty list. The opposite (throw for any empty list) sounds a bit too strict, particularly on LLM use case, as it's easy to mix some empty list in huge semi-structured dataset like scraped documents.

harupy replied:
> infer_signature([]) gives ColSpec(string)

This does seem strange. Could be an uncaught bug.

B-Step62 replied:
I think it's because we use `all()` here, which returns True for empty list.
https://github.com/mlflow/mlflow/blob/v2.7.1/mlflow/types/utils.py#L150

harupy replied:
@B-Step62 Thanks for the investigation. I think `infer_signature([])` should throw.

B-Step62 replied:
@harupy Agreed, should we fix on master branch? It behaves fine in this branch (throws for []) so probably no need for now as long as we merge this soon.

harupy replied:
If it behaves fine, we don't need to.

B-Step62 commented on mlflow/types/utils.py:
```diff
@@ -115,7 +115,7 @@ def _infer_colspec_type(data: Any) -> Union[DataType, Array, Object]:
             properties.append(Property(name=k, dtype=_infer_colspec_type(v)))
         return Object(properties=properties)
-    if isinstance(data, list):
+    if isinstance(data, (list, np.ndarray)):
```
required for support numpy array

harupy commented on tests/models/test_utils.py:
```diff
@@ -321,36 +321,94 @@ def test_enforce_property_with_errors():
+@pytest.mark.parametrize(
+    ("data", "schema"),
+    [
+        # 1. Simply array
+        (["some_sentence1", "some_sentence2"], Array(DataType.string)),
+        # 2. Numpy
```
```suggestion
        # 1. 1D array
        (["some_sentence1", "some_sentence2"], Array(DataType.string)),
        # 2. Numpy 1D array
```

harupy commented on tests/models/test_utils.py:
```diff
@@ -321,36 +321,94 @@ def test_enforce_property_with_errors():
+                )
+            ),
+        ),
+        # 4. Nested array
+        (
+            [
+                [["a", "b"], ["c", "d"]],
+                [["e", "f", "g"], ["h"]],
+                [["g"]],
+            ],
+            Array(Array(Array(DataType.string))),
+        ),
```
@serena-ruan Is there a limit for nesting depth?

serena-ruan replied:
Nope

serena-ruan replied:
Should we add one?

B-Step62 replied:
Practically I don't think someone inputs 10x nested list, but good to have such guardrail. Probably we can put some handy limit like ~5 and raise it once some user hits that (and report)?

harupy replied:
How about 2? We can increase it if there is a feature request.

B-Step62 replied:
That's fine too. We have TensorSpec for use cases with high dimensional array like deep learning right?

B-Step62 committed:
Remove invalid test case

np.array is now valid column value as we added Array type. This test case is covered in test_schema_inference_on_lists.

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
Allow empty array in _enforce_array

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
typo

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

harupy commented on mlflow/models/utils.py:
```diff
@@ -906,15 +906,15 @@ def _enforce_datatype(data: Any, dtype: DataType):
-            data = [data]
-        else:
-            raise MlflowException(f"Expected data to be list, got {type(data).__name__}")
+        raise MlflowException(f"Expected data to be list, got {type(data).__name__}")
```
```suggestion
        raise MlflowException(f"Expected data to be list or numpy array, got {type(data).__name__}")
```

harupy reviewed (approved):
LGTM!

serena-ruan reviewed (approved):
LGTM!
